### ISSUE TITLE
Error types do not impl `From<Infallible>`

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/hex-conservative#128]
We do this in `rust-bitcoin`. Should we be doing it here?

**Comment by apoelstra** [comment]
Yeah, we should. Though it's pretty irritating that the language doesn't do this for us. There should be an implicit conversion from uninhabited types (which the compiler can easily identify) into every other type.

**Comment by tcharding** [comment]
One more thing that should be in the should-be-part-of-rust crate.

**Comment by tcharding** [comment]
Note we removed the macro in `rust-bitcoin` and elected to inline the code. Just add this for each error

```rust
impl From<Infallible> for FooError {
    fn from(never: Infallible) -> Self { match never {} }
}
```

**Comment by tcharding** [comment]
Note this is only for enum errors that have inner error types not struct errors that have data.

**Comment by apoelstra** [comment]
Why not other errors?

**Comment by tcharding** [comment]
I don't know why I/we did it like that. I also just went to try and understand exactly _why_ we do these conversions in the first place and I _still_ don't get it.

I re-read https://github.com/rust-bitcoin/rust-bitcoin/issues/1222#issuecomment-1746186060

Please either put me out of my misery and tell me why these impls are useful or, easier, just tell me if we should add the imples for _all_ errors.

**Comment by apoelstra** [comment]
There is a single unique function from the empty type to any other type. A sane language would provide this conversion for you, but Rust does not (even as a blanket impl in the stdlib) because it has some bad interactions with unsafe code. (In safe Rust it is impossible to construct an uninhabited type, but in unsafe Rust you can do it, sometimes by accident, and if you then try to use it the result is UB.)

Ok, so if we want this conversion, which should exist for mathematical purity reasons, we have to implement it ourselves on every one of our types.

**For error types specifically** though there is a better reason than mathematical purity: it lets somebody call a function which returns a `Result<T, !>` (which is equivalent to `T` but may need to be expressed as a Result because of API constraints) convert this to a `Result<T, SomeRustBitcoinError>` using the `?` operator.

If we didn't do this, then the user would have a harder time using `Result<T, !>` alongside our crate, which would be a tragedy and our fault.

**Comment by apoelstra** [comment]
>  // This looks nice even though the `?` hides the fact that this never errors.

Yeah, it's not great, but it may make sense in generic code. In recent rustcs you can often write

```rust
let Ok(res) = some_function(a);
```
when you know that the error type is a concrete `!`.

> // Could equally be written <diverging path which actually can't be taken>

This is horrific. If you know the type is `!` then you can just destructure the happy path with `let Ok`, or using match in earlier compilers. Then the function doesn't appear to be partial when it's really total. (And if you don't know that it's `!` then you have an actual panicking path, which also sucks.)

**Comment by tcharding** [comment]
Cool, thanks.

**Comment by Kixunil** [comment]
> Though it's pretty irritating that the language doesn't do this for us. There should be an implicit conversion from uninhabited types (which the compiler can easily identify) into every other type.

I was reading a bunch on this and it's surprisingly complicated change in the internals of the compiler. That being said, there was some significant progress on new type checker recently which might unlock it ~soon. Still our MSRV dictates we wait.

> // This looks nice even though the `?` hides the fact that this never errors.

Note that you're assuming the code is not generic.  You can have something like this:

```rust
fn foo<T>(bar: T) -> Result<ReturnType, MyError> where T: Trait, MyError: From<<T as Trait>::Error> {
    // ? is the reason for conversion bound
    let baz = bar.method()?;
    process_baz(baz)
}
```

Here you have to use the `?` operator because of generic and the fact that `Trait::method` is infallible happens to be the property of `T` that the caller passes. Not having the conversion would be annoying. I remember this coming up in practice a few times, I just don't remember where specifically. Maybe it had something to do with reading from slices being infallible in `no_std` IO impls.

> This is horrific.

Yes, it is. There are unstable methods to do what `unwrap`/`unwrap_err` does but with static guarantee of not panicking. IIRC they are called `into_ok` and `into_err`.

**Comment by tcharding** [comment]
I edited your issue @Kixunil and linked back to this so in a few months when I forget this I can find it for `rust-bitcoin` repo.
